### PR TITLE
Change MultiQuery Prompt, Add Hybrid Search (BM25 + Embedding), Cohere Reranker & LLM Chain Filter

### PR DESCRIPTION
**Pull request by davidgxue**
### Description
- Changed MultiQueryRetriever's prompt and parameter so we now keep user's original query + 2 reworded queries (whereas 3 different reworded queries)
- Added Hybrid Search with alpha = 0.5 (equal weight between BM25/TF-TID & embedding model) and retrieving 100 documents for each prompt
- Added Cohere Reranker (keeping 10 documents) after documents are combined from the 3 prompts from multi-query retriever
- Added LLM Chain Filter (using GPT 3.5) to double check each remaining 10 documents is relevant to the question asked
    - This essentially uses GPT 3.5 to ask if document X is relevant to the question (with a YES/NO) response
    - Added our custom boolean parser as the default one in LangChain can be too strict causing error raising in specific cases 

### Technical Changes
- Added api/ask_astro/chains/custom_llm_filter_prompt.py because the default implementation of boolean parser in LangChain library enforces that the GPT-3.5 used must either output "YES" or "NO" or else it raises an error. I essentially took their boolean parser from [here](https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/output_parsers/boolean.py) and re-implemented such that 1. we only check YES/NO is CONTAINED in the response (as sometimes LLM can output "YES.") 2. If neither, then don't raise an error but just return True (in our use case we just don't filter out/still consider this doc relevant).
- Main changes are in api/ask_astro/chains/answer_question.py
- Added cohere package in api/pyproject.toml

### Notes
- Need further more extensive evaluation and comparison on a full Q&A dataset to determine improvements (and potential regressions).
- My initial test/evaluation of the same 74 questions that my teammates have used in the past can be found on [langsmith here](https://smith.langchain.com/o/4942ae8b-e5be-4964-bb1f-7544886fa06d/projects/p/26697fd3-75a4-4056-8fb4-9a1b10420dec?columnVisibilityModel=%7B%22name%22%3Afalse%2C%22start_time%22%3Afalse%2C%22total_tokens%22%3Afalse%2C%22inputs%22%3Atrue%2C%22outputs%22%3Atrue%2C%22tags%22%3Afalse%2C%22in_dataset%22%3Afalse%2C%22last_queued_at%22%3Afalse%2C%22first_token_time%22%3Afalse%2C%22metadata%22%3Afalse%2C%22feedback_stats%22%3Afalse%7D)

### CODE REVIEW

File: api/ask_astro/chains/custom_llm_filter_prompt.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+from langchain.retrievers.document_compressors.chain_filter_prompt import (
+    prompt_template,
+)
+from langchain_core.output_parsers import BaseOutputParser
+from langchain_core.prompts import PromptTemplate
+
+
+class CustomBooleanOutputParser(BaseOutputParser[bool]):
```
Comment (davidgxue) [inline api/ask_astro/chains/custom_llm_filter_prompt.py:8]: Note: this is a changed implementation from langchain. The original code looked like this https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/output_parsers/boolean.py

**Review by sunank200 — COMMENTED** [review]
@vatsrahul1001 did you get a chance to test this? which weaviate index should be used @davidgxue ?

**Comment by vatsrahul1001** [comment]
> @vatsrahul1001 did you get a chance to test this? which weaviate index should be used @davidgxue ?

@sunank200 no I have not tested this end to end, Also as per Steven's [comment](https://astronomer.slack.com/archives/C06CGUXM0LE/p1704489480250459) readthedocs stuff is withdrawn?

**Comment by sunank200** [comment]
> > @vatsrahul1001 did you get a chance to test this? which weaviate index should be used @davidgxue ?
> 
> @sunank200 no I have not tested this end to end, Also as per Steven's [comment](https://astronomer.slack.com/archives/C06CGUXM0LE/p1704489480250459)can you readthedocs stuff is withdrawn?

@vatsrahul1001 yes. The readthedocs from astro-sdk is not there in database now

**Comment by vatsrahul1001** [comment]
> > > @vatsrahul1001 did you get a chance to test this? which weaviate index should be used @davidgxue ?
> > 
> > 
> > @sunank200 no I have not tested this end to end, Also as per Steven's [comment](https://astronomer.slack.com/archives/C06CGUXM0LE/p1704489480250459)can you readthedocs stuff is withdrawn?
> 
> @vatsrahul1001 yes. The readthedocs from astro-sdk is not there in database now

Ok, I will test it tomorrow then

**Comment by davidgxue** [comment]
Yes, I checked with Rahul and he will get onto checking the response quality today!

**Comment by vatsrahul1001** [comment]
> Yes, I checked with Rahul and he will get onto checking the response quality today!


@davidgxue, I have completed the testing and observed an overall improvement in the quality of responses. Even for basic Astro SDK questions, responses have improved, even without the Astro SDK docs. However, for questions in Ask Astro that were specifically designed from the docs, the responses have degraded, which is as expected.

[Results](https://docs.google.com/spreadsheets/d/13cVqNikix82YjCPA4t0XaULg3XccBnvrQUmQa9VwgC0/edit#gid=70777171)

**Review by sunank200 — APPROVED** [review]
LGTM

**Pull request merged**